#' annotate_methylome()
#' This function annotates the methylome with different genomic features (e.g. genes, exon, promoter, etc.)
#'
#' @param mrobj A methylRaw object or a methylRawList object that is generated by mcalls2mkobj()
#' @param genome_info A list of GRanges objects that contains genome infomation.
#' @param save_output_data If specified other than the default "TRUE", then results
#'   are saved in Rda file that is specified by @param outputdata; otherwise either no Rda file is generated.
#' @param outputdata If specified, then the result is saved in the specified file name; otherwise the result
#'   is saved in '$wd/annotate_methylome_data.Rda'
#'
#' @return A table ?that stores the methylation information and annotate_methylome
#'
#' @importFrom methylKit percMethylation unite
#' @importFrom genomation annotateWithFeature
#' @importFrom utils write.table
#'
#' @examples
#'   mydatf <- system.file("extdata","Am.dat",package="BWASPR")
#'   myparf <- system.file("extdata","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(datafile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$datafiles)
#'   ginfo <- get_genome_annotation(myfiles$parameters)
#'   annotate_methylome(AmHE, ginfo, save_output_data=TRUE, outputdata='AmHE_annotate_methylome.Rda')
#'
#' @export


annotate_methylome <- function(mrobj,
                               genome_info,
                               save_output_data = TRUE,
                               outputdata = 'annotate_methylome_data.txt'
                              ){
    # unite all the methlRawList object into a methylBase object and calculate the percentage methylation scores
    #
    meth      <- unite(mrobj, destrand = TRUE)
    perc_meth <- percMethylation(meth, rowids = FALSE, save.txt = FALSE)
    # annotate methylome with generic features from genome information
    #
    for (i in names(genome_info)){
        assign(paste(i, 'annot', sep = '_'),
               annotateWithFeature(as(meth, 'GRanges'),
                                   genome_info[[i]],
                                   strand = TRUE,
                                   intersect.chr = FALSE)
              )
    }
    # add the methylome annotation to the methylBase object
    #
    for (i in names(genome_info)){
        meth[i] <- get(paste(i, 'annot', sep = '_'))@members
    }
    # change the methylBase object into a dataframe and set the correct col names
    #
	meth_data  <- getData(meth)
	label      <- subset(meth_data,
	                     select = c('chr', 'start', 'end', 'strand', names(genome_info))
	                    )
  	annotate_methylome <- cbind(perc_meth, label)

    if (save_output_data == TRUE){
        write.table(annotate_methylome,
                    file = outputdata
                    sep = '\t',
                    row.names = FALSE
                   )
    }
    return(annotate_methylome)
}
