#' explore_dmsg() 
#' This function sorts genes by meth_level per 10kb for individual samples, 
#' and sorts genes by meth_diff per 10kb for each comparison
#'
#' @param mrobj A methyRaw/methRawList object or a methyRawList object
#' @param genome_ann Genome annotation returned by get_genome_annotation()
#' @param show_dmsg a list of dataframe contains dmsg generated by show_dmsg()
#' 
#' @return MISSING  
#' 
#' @importFrom methylKit reorganize unite
#' @importFrom GenomicRanges findOverlaps
#' @importFrom utils write.table 
#' @importFrom S4Vectors subjectHits queryHits
#' @importFrom dplyr group_by %>% summarise
#'
#' @examples
#'   mydatf <- system.file("extdata","Am.dat",package="BWASPR")
#'   myparf <- system.file("extdata","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(datafile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$datafiles,species="Am",study="HE",
#'                        type="CpGhsm",mincov=1,assembly="Amel-4.5")
#'   genome_ann <- get_genome_annotation(myfiles$parameters)
#'   meth_diff <- det_dmsg(AmHE,genome_ann,
#'                         threshold=25.0,qvalue=0.01,mc.cores=4,
#'                         outfile1="AmHE-dmsites.txt", 
#'                         outfile2="AmHE-dmgenes.txt")
#'   show_dmsg <- show_dmsg(AmHE,meth_diff)
#'   explore_dmsg(AmHE,genome_ann,show_dmsg)
#'
#' @export

explore_dmsg <- function(mrobj,genome_ann,show_dmsg){
    message('... explore_dmsg ...')
    # read basic information
    #
    sample_list     <- getSampleID(mrobj)
    gene.gr         <- genome_ann$gene
    comparison_list <- names(show_dmsg)
    # for each sample, return a list of genes sorted by meth level per 10kb
    #
    message('   ... explore individual sample ...')
    lapply(sample_list,function(sample){
        message(paste('      ... explore ',sample,' ...',sep=''))
    	# subset the mrobj
    	#
    	data              <- reorganize(mrobj,
 			                sample.ids=list(sample),
    				        treatment=c(0))[[1]]
    	data.gr           <- as(data,'GRanges')
    	# calc the methylation level in percentage
    	#
    	data.gr$perc_meth <- data.gr$numCs/data.gr$coverage
    	# annotate the msites with genes
    	#
    	match             <- findOverlaps(data.gr,gene.gr)
    	data.gr           <- data.gr[queryHits(match)]
    	gene.gr           <- gene.gr[subjectHits(match)]
        # combine
    	data.df           <- as.data.frame(data.gr)
    	gene.df           <- as.data.frame(gene.gr)
        colnames(gene.df) <- lapply(colnames(gene.df),
    				    function(i) paste('gene',i,sep='_'))
    	data_gene_comb    <- cbind(data.df,gene.df)
    	# calc a set of parameters for each gene
    	#
    	summary <- data_gene_comb %>% group_by(gene_ID) %>%
    	    summarise(n_msites= n(),
    		      n_msites_10kb=round(n()/mean(gene_width)*10000,2),
    		      mlevel_sum=round(abs(sum(perc_meth)),2),
    		      mlevel_n_msites=round(abs(sum(perc_meth))/n(),2),
    		      mlevel_10kb=round(abs(sum(perc_meth))/mean(gene_width)*10000,2))
	# order the genes by mlevel_10kb
	#
	summary <- summary[order(- summary$mlevel_10kb),]
    	write.table(summary,
                    paste(sample,'explore_dmsg.txt',sep='_'),sep='\t')
	})
    # for each comparison, return a list of genes sorted by meth_diff per 10kb
    #
    message('   ... explore comparisons ...')
    lapply(comparison_list, function(comparison){
        message(paste('      ... explore ',comparison,'...',sep=''))
        data            <- show_dmsg[[comparison]]
        comparison      <- toString(comparison)
        sample1         <- unlist(strsplit(comparison,'\\.'))[1]
        sample2         <- unlist(strsplit(comparison,'\\.'))[3]
        data['sample1'] <- data[sample1]
        data['sample2'] <- data[sample2]
	# calc a set of parameters for each gene
        summary <- data %>% group_by(gene_ID) %>%
            summarise(n_msites=n(),
                      n_msites_10kb=round(n()/mean(gene_width)*10000,2),
                      n_dmsites=sum(is.dm),
                      n_dmsites_10kb=round(sum(is.dm)/mean(gene_width)*10000,2),
                      n_dm_n_msites=round(sum(is.dm)/n(),2),
                      diff_meth_n_msites=round(abs(sum(sample1)-sum(sample2))/n(),2),
		      diff_meth_10kb=round(abs(sum(sample1)-sum(sample2))/mean(gene_width)*10000,2))
	# order the genes by diff_meth_10kb
        summary <- summary[order(- summary$diff_meth_10kb),]
        write.table(summary, 
                    paste(comparison,'explore_dmsg.txt',sep='_'),sep='\t')
    })
    message('... explore_dmsg finished ...')
}
