#' explore_dmsg() 
#' This function sorts genes by meth_level per 10kb for individual samples, 
#' and sorts genes by meth_diff per 10kb for each comparison
#'
#' @param mrobj A methyRaw/methRawList object or a methyRawList object
#' @param genome_ann Genome annotation returned by get_genome_annotation()
#' @param show_dmsg a list of sitesframe contains dmsg generated by show_dmsg()
#' 
#' @return MISSING  
#' 
#' @importFrom methylKit reorganize unite
#' @importFrom GenomicRanges findOverlaps
#' @importFrom utils write.table 
#' @importFrom S4Vectors subjectHits queryHits
#' @importFrom dplyr group_by %>% summarise
#'
#' @examples
#'   mydatf <- system.file("extsites","Am.dat",package="BWASPR")
#'   myparf <- system.file("extsites","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(sitesfile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$sitesfiles,species="Am",study="HE",
#'                        type="CpGhsm",mincov=1,assembly="Amel-4.5")
#'   genome_ann <- get_genome_annotation(myfiles$parameters)
#'   dmsgList <- det_dmsg(AmHE,genome_ann,
#'                        threshold=25.0,qvalue=0.01,mc.cores=4,
#'                        destrand=TRUE,
#'                        outfile1="AmHE-dmsites.txt", 
#'                        outfile2="AmHE-dmgenes.txt")
#'   dmgprp <- show_dmsg(AmHE,meth_diff,outflabel="Am_HE")
#'   explore_dmsg(AmHE,genome_ann,dmgprp,outflabel="Am_HE")
#'
#' @export


explore_dmsg <- function(mrobj,genome_ann,dmgprp,outflabel="") {
    message('... explore_dmsg ...')
    # read basic information
    #
    sample_list     <- getSampleID(mrobj)
    gene.gr         <- genome_ann$gene
    comparison_list <- names(dmgprp)
    # for each sample, return a list of genes sorted by meth level per 10kb
    #
    message('   ... explore individual sample ...')
    lapply(sample_list, function(sample) {
        message(paste('      ... explore ',sample,' ...',sep=''))
        # subset the mrobj
        #
        sites             <- reorganize(mrobj,
                                        sample.ids=list(sample),
                                        treatment=c(0))[[1]]
        sites.gr          <- as(sites,'GRanges')
        # calc the methylation level in percentage
        #
        sites.gr$perc_meth <- (sites.gr$numCs/sites.gr$coverage) * 100
        # annotate the msites with genes
        #
        match             <- findOverlaps(sites.gr,gene.gr)
        sites.gr          <- sites.gr[queryHits(match)]
        gene.gr           <- gene.gr[subjectHits(match)]
        # combine
        sites.df          <- as.data.frame(sites.gr)
        gene.df           <- as.data.frame(gene.gr)
        colnames(gene.df) <- lapply(colnames(gene.df),
                                    function(i) paste('gene',i,sep='_'))
        sites_gene        <- cbind(sites.df,gene.df)
        # calc a set of parameters for each gene
        #
        summary <- sites_gene %>% group_by(gene_ID) %>%
            summarise(glink = paste("https://www.ncbi.nlm.nih.gov/gene/?term",
                                    unique(gene_Name),sep="="),
                      gwidth = round(mean(gene_width),2),
                      nbrsites = n(),
                      nbrper10kb = round((nbrsites/gwidth)*10000,2),
                      pmsum = round(sum(perc_meth),2),
                      pmpersite = round(pmsum/nbrsites,2),
                      pmpernucl = round(pmsum/gwidth,2)
                     )
        # order the genes by pmpernucl
        #
        summary <- summary[order(- summary$pmpernucl),]
        summary <- subset(summary, select = -c(pmsum))
        colnames(summary) <- c("gene_ID","gene_link","gwidth",
                               "#Sites","#per10Kb","%perSite","%pNucl")
        outfile <- paste("ogl",outflabel,sep="-")
        outfile <- paste(outfile,sample,sep="_")
        wtoutfile <- paste(outfile,"txt",sep=".")
        write.table(summary, wtoutfile, sep='\t',
                    row.names=FALSE, quote=FALSE)
    })

    # for each comparison, return a list of genes sorted by meth_diff per 10kb
    #
    message('   ... explore comparisons ...')
    lapply(comparison_list, function(comparison) {
        message(paste('      ... explore ',comparison,'...',sep=''))
        sites            <- dmgprp[[comparison]]
        comparison       <- toString(comparison)
        sample1          <- unlist(strsplit(comparison,'\\.'))[1]
        sample2          <- unlist(strsplit(comparison,'\\.'))[3]
        sites['sample1'] <- sites[sample1]
        sites['sample2'] <- sites[sample2]
        # calc a set of parameters for each gene
        summary <- sites %>% group_by(gene_ID) %>%
            summarise(glink = paste("https://www.ncbi.nlm.nih.gov/gene/?term",
                                    unique(gene_Name),sep="="),
                      gwidth = round(mean(gene_width),2),
                      nbrsites = n(),
                      nbrper10kb = round((nbrsites/gwidth)*10000,2),
                      nbrdmsites = sum(is.dm),
                      nbrdmper10kb = round((nbrdmsites/gwidth)*10000,2),
                      prcntdm = round((nbrdmsites/nbrsites)*100,2),
                      admpersite = round(abs(sum(sample1)-sum(sample2))/nbrsites,2),
                      admpernucl = round(abs(sum(sample1)-sum(sample2))/gwidth,2)
                     )
        # order the genes by admpernucl
        summary <- summary[order(- summary$admpernucl),]
        colnames(summary) <- c("gene_ID","gene_link","gwidth","#Sites","#per10Kb",
                               "#dmSites","#dmsp10kb","%dmSites","ADMpSite","ADMpNucl")
        outfile <- paste("ogl",outflabel,sep="-")
        outfile <- paste(outfile,comparison,sep="_")
        wtoutfile <- paste(outfile,"txt",sep=".")
        write.table(summary, wtoutfile, sep='\t', row.names=FALSE, quote=FALSE)
    })
    message('... explore_dmsg finished ...')
}
