#' plot_genome_annotation()
#' This function plots a representative figure of genome annotation
#'
#' @param genome_annotation A table generated by genome_annotation()
#'   ?that stores the methylation information and genome_annotation
#' @param mrobj A methylRaw object or a methylRawList object.
#' @param plot TRUE|FALSE. If TRUE, plot the heatmap figure, if FALSE, save the heatmap figure.
#' @param output_file ?An option to specify the dir/file names for output figure.
#' @param number_of_sites Number of methylated sites that are plotted on figure
#'
#' @return A data frame that store the genome annotation data
#'
#  @import methods utils
#' @import ComplexHeatmap
#' @importFrom ComplexHeatmap draw
#' @import circlize
#' @importFrom methylKit getSampleID
#' @importFrom grid unit
#'
#' @examples
#'   mydatf <- system.file("extdata","Am.dat",package="BWASPR")
#'   myparf <- system.file("extdata","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(datafile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$datafiles)
#'   genome <- read_genome_info(myfiles$parameters)
#'   genome_annotation_info <- genome_annotation(AmHE, genome, save_output_data = FALSE)
#'   plot_genome_annotation(AmHE, genome_annotation_info)
#'
#' @export

plot_genome_annotation <- function(mrobj,
                                   genome_annotation_info,
                                   plot = FALSE,
                                   output_file = 'genome_annotation.pdf',
                                   number_of_sites = 200){

    # import the library for this function
    library(ComplexHeatmap)
    library(circlize)
    # prep the data to plot in the figure
    #
    plot_matrix <- head(genome_annotation_info, number_of_sites)
    label_matrix = as.data.frame(ifelse(plot_matrix == 0, "No", "Yes"))
    sample_list <- getSampleID(mrobj)
    # annoation for the heatmap
    #
    ha = HeatmapAnnotation(df = data.frame(caste = sample_list))
    ha_gene = rowAnnotation(df = label_matrix['gene'],
                            name = 'gene',
                            show_annotation_name = TRUE,
                            width = unit(2, "mm"),
                            col = list(gene = c('No' =  "gray", 'Yes' = "blue")),
                            annotation_name_side = 'top'
                            )
    ha_exon = rowAnnotation(df = label_matrix['exon'],
                            name = 'exon',
                            show_annotation_name = TRUE,
                            width = unit(2, "mm"),
                            col = list(exon = c('No' =  "gray", 'Yes' = "blue")),
                            annotation_name_side = 'top'
                            )
    ha_pcexon = rowAnnotation(df = label_matrix['pcexon'],
                              name = 'pcexon',
                              show_annotation_name = TRUE,
                              width = unit(2, "mm"),
                              col = list(pcexon = c('No' =  "gray", 'Yes' = "blue")),
                              annotation_name_side = 'top'
                              )
    ha_cds = rowAnnotation(df = label_matrix['CDS'],
                           name = 'CDS',
                           show_annotation_name = TRUE,
                           width = unit(2, "mm"),
                           col = list(CDS = c('No' =  "gray", 'Yes' = "blue")),
                           annotation_name_side = 'top'
                           )
    ha_promoter = rowAnnotation(df = label_matrix['promoter'],
                                name = 'promoter',
                                show_annotation_name = TRUE,
                                width = unit(2, "mm"),
                                col = list(promoter = c('No' =  "gray", 'Yes' = "blue")),
                                annotation_name_side = 'top'
                                )

    heatmap <- Heatmap(plot_matrix[, 1:length(sample_list)],
                       col = colorRamp2(c(0, 50, 100), c('blue4', 'blue', 'lightskyblue')),
                       name = 'methylation level',
                       show_row_names = FALSE,
                       column_title = 'Representative methylome annotation', row_title = 'methylated sites',
                       top_annotation = ha, top_annotation_height = unit(3, "mm")
                      )
    # if not plot, then only draw on plots,
    # else, save the figure into a file directory that is specified by user or default
    #
    if (plot == FALSE){
        draw(heatmap + ha_gene + ha_exon + ha_pcexon + ha_cds + ha_promoter)
    } else{
        pdf(output_file, length(sample_list) * 3, number_of_sites * 0.03)
        draw(heatmap + ha_gene + ha_exon + ha_pcexon + ha_cds + ha_promoter)
        dev.off()
    }
}
