#' plot_genome_annotation()
#' This function plots a heatmap to show the annotation of the first 200 methylated sites
#'
#' @param mrobj A methylRaw object or a methylRawList object.
#' @param methylome_annotation A table generated by annotate_methylome()
#'   that stores the methylome information and methylome annotation information
#' @param plot TRUE|FALSE. If TRUE, plot the heatmap figure, if FALSE, save the heatmap figure.
#' @param output_file An option to specify the dir/file names for output figure.
#'
#' @return A data frame that store the genome annotation data
#'
#' @import methods utils
#' @import ComplexHeatmap
#' @importFrom ComplexHeatmap draw
#' @import circlize
#' @importFrom methylKit getSampleID
#' @importFrom grid unit
#'
#' @examples
#'   mydatf <- system.file("extdata","Am.dat",package="BWASPR")
#'   myparf <- system.file("extdata","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(datafile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$datafiles)
#'   genome <- get_genome_annotation(myfiles$parameters)
#'   methylome_annotation <- annotate_methylome(AmHE, genome, save_output_data = FALSE)
#'   plot_genome_annotation(AmHE, methylome_annotation)
#'
#' @export

plot_genome_annotation <- function(mrobj,
                                   methylome_annotation,
                                   plot = FALSE,
                                   output_file = 'methylome_annotation.pdf'){
    # number of methylated sites that are shown on the heatmap
    number_of_sites = 200
    # prep the dataframe to plot in the figure
    plot_matrix              <- head(methylome_annotation, number_of_sites)
    label_matrix             <- as.data.frame(plot_matrix)
    # prep the label for intergenic region, gene-exon, gene-intron
    label_matrix['gene']     <- label_matrix['gene'] + label_matrix['exon']
    label_matrix['gene'][label_matrix['gene'] == 0] <- 'intergenic'
    label_matrix['gene'][label_matrix['gene'] == 1] <- 'gene:intron'
    label_matrix['gene'][label_matrix['gene'] == 2] <- 'gene:exon'
    # prep the label for CDS and promoter
    label_matrix['CDS']      <- ifelse(plot_matrix['CDS'] == 0, 'No', 'Yes')
    label_matrix['promoter'] <- ifelse(plot_matrix['promoter'] == 0, 'No', 'Yes')
    sample_list              <- getSampleID(mrobj)
    # annoation for the heatmap
    #
    ha = HeatmapAnnotation(df = data.frame(caste = sample_list))
    ha_gene = rowAnnotation(df = label_matrix['gene'],
                            name = 'gene',
                            show_annotation_name = TRUE,
                            width = unit(2, 'mm'),
                            col = list(gene = c('intergenic' = 'gray', 'gene:exon' = 'blue', 'gene:intron' = 'red')),
                            annotation_name_side = 'top'
                            )
    ha_cds = rowAnnotation(df = label_matrix['CDS'],
                           name = 'CDS',
                           show_annotation_name = TRUE,
                           width = unit(2, 'mm'),
                           col = list(CDS = c('No'= 'gray', 'Yes' = 'blue')),
                           annotation_name_side = 'top'
                           )
    ha_promoter = rowAnnotation(df = label_matrix['promoter'],
                                name = 'promoter',
                                show_annotation_name = TRUE,
                                width = unit(2, 'mm'),
                                col = list(promoter = c('No' = 'gray', 'Yes' = 'blue')),
                                annotation_name_side = 'top'
                                )

    heatmap <- Heatmap(plot_matrix[, 1:length(sample_list)],
                       col = colorRamp2(c(0, 50, 100), c('blue4', 'blue', 'lightskyblue')),
                       name = 'methylation level',
                       show_row_dend = TRUE,
                       row_dend_reorder = FALSE,
                       show_row_names = FALSE,
                       column_title = 'Representative methylome annotation', row_title = 'methylated sites',
                       top_annotation = ha, top_annotation_height = unit(3, "mm")
                      )
    # if not plot, then only draw on plots,
    # else, save the figure into a file directory that is specified by user or default
    #
    if (plot == FALSE){
        draw(heatmap + ha_gene + ha_cds + ha_promoter)
    } else{
        pdf(output_file, length(sample_list) * 3, number_of_sites * 0.03)
        draw(heatmap + ha_gene + ha_cds + ha_promoter)
        dev.off()
    }
}
