#' show_dmsg() 
#' This function will plot heatmap of methylation level of isites in dmgenes
#'
#' @param mrobj A methyRaw object or a methyRawList object
#' @param dmgenes.file A tab delimited file stores genes with differential methylated sites
#' @param min.sites minimal number of msites per genes
#' @param query.genes A list of gene IDs 
#' 
#' @return
#' 
#' @importFrom methylKit percMethylation reorganize unite
#' @importFrom GenomicRanges findOverlaps
#' @importFrom gplots heatmap.2 greenred
#' @importFrom utils write.table capture.output
#' @importFrom S4Vectors subjectHits queryHits
#' @importFrom dplyr group_by
#'
#' @examples
#'   mydatf <- system.file("extdata","Am.dat",package="BWASPR")
#'   myparf <- system.file("extdata","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(datafile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$datafiles,species="Am",study="HE",
#'                        type="CpGhsm" mincov=1,assembly="Amel-4.5")
#'   genome_ann <- get_genome_annotation(myfiles$parameters)
#'   meth_diff <- det_dmsg(AmHE,genome_ann,
#'                         threshold=0.25,qvalue=0.05,mc.cores=4,
#'                         outfile1="AmHE-dmsites.txt", 
#'                         outfile2="AmHE-dmgenes.txt")
#'   show_dmsg(AmHE,'AmHE-dmgenes.txt')
#'
#' @export

show_dmsg <- function(mrobj,dmgenes.file,min.nsites=2,query.genes=''){
    message('... show_dmsg() ...')
    # read dmgenes file generated by det_dmsg()
    #
    dmgenes <- read.table(dmgenes.file,
                          header=TRUE,
                          fill=TRUE,
                          sep='\t')
    # create a list contains all comparisons
    #
    sample_match_list <- unique(dmgenes[,'comparison'])
    sample_match_list <- sample_match_list[grep('.vs.',sample_match_list)]
    # for each comparison
    # 
    lapply(sample_match_list, function(sample_match){
        # parse the 2 samples in this comparison
        #
        sample_match    <- toString(sample_match)
        sample1         <- unlist(strsplit(sample_match,'\\.'))[1]
        sample2         <- unlist(strsplit(sample_match,'\\.'))[3]
        pair_dmgenes.gr <- as(dmgenes[dmgenes['comparison'] == sample_match,],
                              'GRanges')
        message(paste('... comparing ',sample1,' & ',sample2,' ...',sep=''))
        # parse the mrobj with interested samples and calc meth level
        #
        pair_mrobj      <- reorganize(mrobj,
                                      sample.id=list(sample1,sample2),
                                      treatment=c(0,1))
        pair_meth       <- unite(pair_mrobj)
        p_meth          <- round(percMethylation(pair_meth,
                                                 rowids=FALSE,
                                                 save.txt=FALSE),2)
        pair_p_meth     <- cbind(pair_meth,p_meth)
        pair_p_meth.gr  <- as(pair_p_meth,'GRanges')
        # match
        #
        match              <- findOverlaps(pair_dmgenes.gr,pair_p_meth.gr)
        sub_pair_p_meth.gr <- pair_p_meth.gr[subjectHits(match)]
        sub_dmgenes.gr     <- pair_dmgenes.gr[queryHits(match)]

        sub_pair_p_meth    <- as.data.frame(sub_pair_p_meth.gr,row.names=NULL)
        sub_dmgenes        <- as.data.frame(sub_dmgenes.gr,row.names=NULL)
        colnames(sub_dmgenes) <- lapply(colnames(sub_dmgenes), 
                                        function(i) paste('gene',i,sep='_'))

        site_dmg <- cbind(sub_pair_p_meth,
                          sub_dmgenes)

        site_dmg <- site_dmg[,c(sample1,sample2,
                                'seqnames','start','end','width','strand',
                                'gene_Dbxref','gene_width','gene_ID')]
                                            
        if (query.genes != ''){
            site_dmg <- site_dmg[site_dmg$gene_ID %in% query.genes,]
        }
        dmgene_sites <- group_by(site_dmg,gene_Dbxref)
        # prep data for ploting and saving
        #
        out <- split(site_dmg,factor(dmgene_sites$gene_Dbxref))
        # plot heatmap for each dmgene
        #
        pdf(paste(sample_match,'.pdf',sep=''))

        for (i in out) {
            if (dim(i)[1] < min.nsites) {
                # pass if dim(i) < c(2, 2)
            }else{
            plot <- i[,c(sample1,sample2)]
            heatmap.2(as.matrix(plot), 
                      margin=c(10,10),
                      dendrogram='none',
                      Rowv=FALSE,
                      col=greenred(10),
                      trace='none',
                      main=paste('msites@',unique(i$gene_Dbxref),sep=''),
                      srtCol=45,
                      RowSideColors=as.character(as.numeric(i$gene_Dbxref)))

            }
        }
        dev.off()
        # save files into tab delimited files                              
        #
        write.table(site_dmg,
                    file=(paste(sample_match,'show_dmsg_df.txt',sep='_')),
                    sep='\t',
                    row.names=FALSE)
        capture.output(out,
                       file=paste(sample_match,'show_dmsg_list.txt',sep='_'))
    }
    )
    message('... show_dmsg() finished ...')
}
