#' annotate_methylome()
#' This function calculate the number of methylated sites and average methylation level for each gene
#'
#' @param mrobj A methylRaw object or a methylRawList object that is generated by mcalls2mkobj()
#' @param genome_info A list of GRanges objects that contains genome infomation.
#' @param outputdata If specified, then the result is saved in the specified file name; otherwise the result
#'   is saved in 'methyl_level_in_gene'
#'
#' @return A table that stores the information of methylation level in each gene
#'
#' @importFrom methylKit percMethylation unite getSampleID
#' @importFrom genomation annotateWithFeature
#' @importFrom utils write.table
#'
#' @examples
#'   mydatf <- system.file("extdata","Am.dat",package="BWASPR")
#'   myparf <- system.file("extdata","Am.par",package="BWASPR")
#'   myfiles <- setup_BWASPR(datafile=mydatf,parfile=myparf)
#'   AmHE <- mcalls2mkobj(myfiles$datafiles)
#'   ginfo <- get_genome_annotation(myfiles$parameters)
#'   cal_methyl_level_in_gene(AmHE, ginfo, outputdata='AmHE_methyl_level_in_gene.txt')
#'
#' @export


cal_methyl_level_in_gene <- function(mrobj,
                                     genome_info,
                                     outputdata = 'methyl_level_in_gene.txt'){
    # unite all the methlRawList object into a methylBase object and calculate the percentage methylation scores
    #
    meth      <- unite(mrobj, destrand = TRUE)
    perc_meth <- percMethylation(meth, rowids = FALSE, save.txt = FALSE)
    sampleid  <- getSampleID(meth)
    ### annotate methylome with different genes, here is just a test, choose the first 50 genes for test
    ##
    # Notice only first 50 genes here!!!!
    gene      <- genome_info$gene[1:50]
    for (i in seq_along(gene)){
        gene_name <- gene[i]$ID
        annot <- annotateWithFeature(as(meth, 'GRanges'), gene[i], strand = TRUE)
        meth[gene_name] <- annot@members
    }
    # change the methylBase object into a dataframe and set the correct col names
    #
    meth_data  <- getData(meth)
    methylome_gene_annotation <- cbind.data.frame(perc_meth, meth_data)
    # calculate the methylaton level in each gene
    methyl_level_in_gene <- data.frame()
    for (i in seq_along(gene)){
        methyl_level_in_gene['number_methyl_sites', gene[i]$ID] <- nrow(methylome_gene_annotation[methylome_gene_annotation[gene[i]$ID] == 1, ])
        for (j in seq_along(sampleid)){
            methyl_level_in_gene[sampleid[j], gene[i]$ID] <- mean(methylome_gene_annotation[methylome_gene_annotation[gene[i]$ID] == 1, j])
        }
    }


    write.table(methyl_level_in_gene,
                file = outputdata,
                sep = '\t',
                row.names = FALSE
               )
    return(methyl_level_in_gene)
}
